### ISSUE TITLE
How does shape define?

### ISSUE THREAD

**Issue by elamaunt** [cbovar/ConvNetSharp#74]
Hello! My question is: why example uses in train function the shape "new Shape(1, 1, 10, 1)"? Why not "new Shape(10)"?

**Comment by cbovar** [comment]
Hi! ConvnetSharp assumes that shapes have this format:  [width, height, class, batch].

It may be possible to make it simple and allow the use of `new Shape(10) `or `new Shape(10, batch) `by modifying a bit the Shape class.

**Comment by elamaunt** [comment]
Thank you for reply!
 Can you help me? Why same code on C# does not work properly? It is XOR realisation.

            var net = new Net<double>();
            net.AddLayer(new InputLayer(1, 1, 2));

            net.AddLayer(new FullyConnLayer(3));
            net.AddLayer(new TanhLayer());

            net.AddLayer(new FullyConnLayer(2));
            net.AddLayer(new SoftmaxLayer(2));
            
            var x0 = new Volume(new double[] { 1, 1 }, new Shape(2));
            var x1 = new Volume(new double[] { 0, 1 }, new Shape(2));
            var x2 = new Volume(new double[] { 1, 0 }, new Shape(2));
            var x3 = new Volume(new double[] { 0, 0 }, new Shape(2));
            
            var trainer = new SgdTrainer<double>(net) { LearningRate = 0.2, L2Decay = 0.001 };
            
            for (int i = 0; i < 2000; i++)
            {
                trainer.Train(x0, new Volume(new double[] { 0.0 }, new Shape(1, 1, 1, 1)));
                trainer.Train(x1, new Volume(new double[] { 1.0 }, new Shape(1, 1, 1, 1)));
                trainer.Train(x2, new Volume(new double[] { 1.0 }, new Shape(1, 1, 1, 1)));
                trainer.Train(x3, new Volume(new double[] { 0.0 }, new Shape(1, 1, 1, 1)));
            }

            Console.WriteLine("x0: " + net.Forward(x0).Get(0)); // result in 1
            Console.WriteLine("x1: " + net.Forward(x1).Get(0)); // result in 1
            Console.WriteLine("x2: " + net.Forward(x2).Get(0)); // result in 1
            Console.WriteLine("x3: " + net.Forward(x3).Get(0)); // result in 1

            Console.ReadLine();

All volumes give 1. I'm slightly at a loss. I thought it was because of Shape.
The same code on js works perfectly. =( I really hope for your help. https://gist.github.com/willguitaradmfar/12f7efb2ab9534755619ba4dce72f988

**Comment by cbovar** [comment]
You are using a **SoftmaxLayer** with two classes to solve this problem. When training the network, for each input you should give a volume that has '1' on the correct class as ground truth (one-hot vector of length 2).

you could try that:
```csharp
var net = new Net<double>();
net.AddLayer(new InputLayer(1, 1, 2));

net.AddLayer(new FullyConnLayer(3));
net.AddLayer(new TanhLayer());

net.AddLayer(new FullyConnLayer(2));
net.AddLayer(new SoftmaxLayer(2));

var x0 = new Volume(new double[] { 1, 1 }, new Shape(2));
var x1 = new Volume(new double[] { 0, 1 }, new Shape(2));
var x2 = new Volume(new double[] { 1, 0 }, new Shape(2));
var x3 = new Volume(new double[] { 0, 0 }, new Shape(2));

var trainer = new SgdTrainer<double>(net) { LearningRate = 0.2, L2Decay = 0.001 };

for (var i = 0; i < 2000; i++)
{
    trainer.Train(x0, new Volume(new[] { 1.0, 0.0 }, new Shape(1, 1, 2, 1)));
    trainer.Train(x1, new Volume(new[] { 0.0, 1.0 }, new Shape(1, 1, 2, 1)));
    trainer.Train(x2, new Volume(new[] { 0.0, 1.0 }, new Shape(1, 1, 2, 1)));
    trainer.Train(x3, new Volume(new[] { 1.0, 0.0 }, new Shape(1, 1, 2, 1)));

    Console.WriteLine(trainer.Loss);
}

Console.WriteLine("x0: " + net.Forward(x0).Get(0)); // result in 1
Console.WriteLine("x1: " + net.Forward(x1).Get(0)); // result in 1
Console.WriteLine("x2: " + net.Forward(x2).Get(0)); // result in 1
Console.WriteLine("x3: " + net.Forward(x3).Get(0)); // result in 1

Console.ReadLine();
```

**Comment by elamaunt** [comment]
Thank you very much, it works!
